File: javaexample/src/main/java/io/chthonic/mythos/javaexample/ui/vus/FusVu.java
package io.chthonic.mythos.javaexample.ui.vus;

import android.app.Activity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.appcompat.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import io.chthonic.mythos.javaexample.R;
import io.chthonic.mythos.javaexample.databinding.ActivityFusBinding;
import io.chthonic.mythos.javaexample.ui.fragments.RoFragment;
import io.chthonic.mythos.javaexample.ui.layouts.DahLayout;
import io.chthonic.mythos.mvp.Vu;

/**
 * Created by jhavatar on 3/3/2016.
 */
public class FusVu extends Vu<ActivityFusBinding> {

    private Runnable toggleRoListener = null;
    private Runnable toggleDahListener = null;

    public FusVu(LayoutInflater layoutInflater,
                 Activity activity,
                 Fragment fragment,
                 ViewGroup parentView) {
        super(layoutInflater, activity, fragment, parentView);
    }

    public Runnable getToggleRoListener() {
        return toggleRoListener;
    }

    public void setToggleRoListener(Runnable toggleRoListener) {
        this.toggleRoListener = toggleRoListener;
    }

    public Runnable getToggleDahListener() {
        return toggleDahListener;
    }

    public void setToggleDahListener(Runnable toggleDahListener) {
        this.toggleDahListener = toggleDahListener;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        getRootView().findViewById(R.id.button_toggle_ro).setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {
                Runnable listener = toggleRoListener;
                if (listener != null) {
                    listener.run();
                }
            }
        });

        getRootView().findViewById(R.id.button_toggle_dah).setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View view) {
                Runnable listener = toggleDahListener;
                if (listener != null) {
                    listener.run();
                }
            }
        });
    }

    public void updateRoDisplay(boolean showRo) {
        FragmentManager fragmentManager = ((AppCompatActivity) getActivity()).getSupportFragmentManager();
        Fragment fragment = fragmentManager.findFragmentByTag(RoFragment.TAG);
        if (showRo && (fragment == null)) {
            fragment = new RoFragment();
            fragmentManager.beginTransaction()
                    .replace(R.id.fragment_container, fragment, RoFragment.TAG)
                    .commit();

        } else if (!showRo && (fragment != null)) {
            fragmentManager.beginTransaction().remove(fragment).commit();
        }
    }

    public void updateDahDisplay(boolean showDah) {
        ViewGroup childLayout = getRootView().findViewById(R.id.child_container);
        View dahLayout = childLayout.findViewById(R.id.dah_layout);
        if (showDah && (dahLayout == null)) {
            dahLayout = new DahLayout(getActivity(), getActivity().getResources().getString(R.string.fus_lifecycle_key));
            dahLayout.setId(R.id.dah_layout);
            childLayout.addView(dahLayout);

        } else if (!showDah && (dahLayout != null)) {
            childLayout.removeView(dahLayout);
        }
    }

    public void setText(String text) {
        ((TextView) getRootView().findViewById(R.id.fus_text)).setText(text);
    }
}

File: kotlinexample/src/main/java/io/chthonic/mythos/kotlinexample/ui/vus/FusVu.kt
package io.chthonic.mythos.kotlinexample.ui.vus

import android.app.Activity
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import io.chthonic.mythos.kotlinexample.R
import io.chthonic.mythos.kotlinexample.databinding.ActivityFusBinding
import io.chthonic.mythos.kotlinexample.ui.fragments.RoFragment
import io.chthonic.mythos.kotlinexample.ui.layouts.DahLayout
import io.chthonic.mythos.mvp.Vu

/**
 * Created by jhavatar on 3/3/2016.
 */
class FusVu(
    layoutInflater: LayoutInflater,
    activity: Activity,
    fragment: Fragment? = null,
    parentView: ViewGroup? = null,
) : Vu<ActivityFusBinding>(
    layoutInflater,
    activity = activity,
    fragment = fragment,
    parentView = parentView,
) {

    var toggleRoListener: (() -> Unit)? = null
    var toggleDahListener: (() -> Unit)? = null

    override fun inflateBinding(
        layoutInflater: LayoutInflater,
        parentView: ViewGroup?,
    ): ActivityFusBinding =
        if (parentView != null) {
            ActivityFusBinding.inflate(layoutInflater, parentView, false)
        } else {
            ActivityFusBinding.inflate(layoutInflater)
        }

    override fun onCreate() {
        super.onCreate()

        binding.buttonToggleRo.setOnClickListener { _ ->
            toggleRoListener?.invoke()
        }

        binding.buttonToggleDah.setOnClickListener { _ ->
            toggleDahListener?.invoke()
        }
    }

    fun updateRoDisplay(show: Boolean) {
        val supportFragmentManager = (activity as AppCompatActivity).supportFragmentManager
        val rohFragment: Fragment? = supportFragmentManager.findFragmentByTag(RoFragment.TAG)

        if (show and (rohFragment == null)) {
            val fragment = RoFragment()
            supportFragmentManager.beginTransaction()
                .replace(R.id.fragment_container, fragment, RoFragment.TAG)
                .commit()

        } else if (!show && (rohFragment != null)) {
            supportFragmentManager.beginTransaction().remove(rohFragment).commit()
        }
    }

    fun updateDahDisplay(show: Boolean) {
        val dah = binding.childContainer.findViewById<DahLayout>(R.id.dah_layout)

        if (show && (dah == null)) {
            val roLayout =
                DahLayout(activity, activity.resources.getString(R.string.fus_lifecycle_key))
            roLayout.id = R.id.dah_layout
            binding.childContainer.addView(roLayout)

        } else if (!show && (dah != null)) {
            (dah.parent as ViewGroup).removeView(dah)
            rootView.invalidate()
        }
    }

    fun setText(text: String) {
        binding.fusText.text = text
    }

}